File: src/Strings/AddBinary.java
package Strings;

public class AddBinary {
    /*
     * The current implementation of the addBinary method converts the binary
     * strings to integers, performs the addition using bitwise operations, and then
     * converts the result back to a binary string. However, this approach may not
     * handle very large binary strings correctly due to integer overflow.
     */
    public static String addBinary(String x, String y) {
        int a = Integer.parseInt(x,2);
        int b = Integer.parseInt(y,2);

        while (b != 0) {
            int carry = (a & b) << 1;
            a = a ^ b;
            b = carry;
        }

        return Integer.toBinaryString(a);
    }

    /*
     * This implementation handles binary addition directly on the strings and
     * avoids potential issues with integer overflow.
     * time complextiy: O(n)
     * space complexity: O(n)
     */
    public static String addBinary_2(String x, String y) {
        StringBuffer result = new StringBuffer();
        int i = x.length() - 1;
        int j = y.length() - 1;
        int carry = 0;

        while(i >= 0 || j >= 0 || carry != 0) {
            int bit1 = (i >= 0) ? x.charAt(i) - '0' : 0;
            int bit2 = (j >= 0) ? y.charAt(i) - '0' : 0;

            int sum = bit1 + bit2 + carry;
            result.append(sum % 2);
            carry = sum /2;
        }

        return result.reverse().toString();
    }
    public static void main(String[] args) {
        String a = "1010";
        String b = "1011";    

        System.out.println(addBinary(a,b));
        System.out.println(addBinary_2(a, b));
    }
    
}

File: src/Strings/AddBinary.py
#
#  * This implementation handles binary addition directly on the strings and
#  * avoids potential issues with integer overflow.
#  * time complexity: O(n)
#  * space complexity: O(n)
#  
def addBinary(s1, s2):
    # Initialize pointers for the end of both strings
    i = len(s1) - 1
    j = len(s2) - 1
    # Initialize carry to handle overflow during addition
    carry = 0
    # Initialize a list to store the result in reverse order
    result = []

    # Loop until all bits are processed or there is a carry left
    while i >= 0 or j >= 0 or carry != 0:
        # Get the current bit from s1, or 0 if out of bounds
        bit1 = int(s1[i]) if i >= 0 else 0
        # Get the current bit from s2, or 0 if out of bounds
        bit2 = int(s2[j]) if j >= 0 else 0

        # Calculate the sum of the two bits and the carry
        bit_sum = bit1 + bit2 + carry
        # Append the least significant bit of the sum to the result
        result.append(str(bit_sum % 2))
        # Update the carry for the next iteration
        carry = bit_sum // 2
        # Move to the previous bit in both strings
        i -= 1
        j -= 1
    
    # Reverse the result list, join it into a string, and remove leading zeros
    result = ''.join(reversed(result)).lstrip('0')
    # Return the result, or '0' if the result is empty
    return result if result else '0'


# Example usage
s1 = '00100'
s2 = '010'
print(addBinary(s1, s2))
